### PR TITLE
Improved half/float16 support for C++ & CUDA targets

### PR DESCRIPTION
**Pull request by juliusikkala**
This is a fixed version of #8208 with improvements from #8985. That PR got just completely wrecked by a local rebase somehow (should configure git pull to not rebase I guess), so I squashed the actual diff and put it here again.

### CODE REVIEW

File: prelude/slang-cuda-prelude.h [hunk 1]
```diff
@@ -1679,9 +1876,9 @@ SLANG_FORCE_INLINE SLANG_CUDA_CALL float F32_rsqrt(float f)
 {
     return ::rsqrtf(f);
 }
-SLANG_FORCE_INLINE SLANG_CUDA_CALL float F32_sign(float f)
```
Comment (kaizhangNV) [inline prelude/slang-cuda-prelude.h:1682]: can you explain why this change is necessary?

File: tests/hlsl-intrinsic/scalar-half.slang [hunk 2]
```diff
@@ -0,0 +1,120 @@
+//TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=CHECK):-cpu -compute -shaderobj
+//DISABLED_TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=CHECK):-slang -compute -shaderobj
+// Disabled due to DXC seemingly mishandling the NaN corner case of min() & max()
```
Comment (kaizhangNV) [inline tests/hlsl-intrinsic/scalar-half.slang:3]: do we have issue tracked this for further investigation?

File: tests/hlsl-intrinsic/scalar-half.slang [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+//DISABLED_TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=CHECK):-slang -compute -dx12 -profile cs_6_2 -use-dxil -Xslang... -Xdxc -Gis -X. -shaderobj
+// `min(half, half)` seems to crash spirv-opt, so disable optimization for now to allow this test to pass.
```
Comment (kaizhangNV) [inline tests/hlsl-intrinsic/scalar-half.slang]: Also, for this do we have issue to track this?

File: tests/hlsl-intrinsic/scalar-half.slang [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+//TEST(compute, vulkan):COMPARE_COMPUTE_EX(filecheck-buffer=CHECK):-vk -compute -emit-spirv-directly -shaderobj -Xslang... -O0 -X.
+//TEST(compute, vulkan):COMPARE_COMPUTE_EX(filecheck-buffer=CHECK):-cuda -compute -shaderobj
+//DISABLED_TEST(compute):COMPARE_COMPUTE_EX(filecheck-buffer=CHECK):-wgpu -compute -shaderobj
```
Comment (kaizhangNV) [inline tests/hlsl-intrinsic/scalar-half.slang:11]: why can't it be supported on webgpu? Is because of webgpu limitation?

File: prelude/slang-cuda-prelude.h (hunk 1 above)
Comment (juliusikkala) [reply prelude/slang-cuda-prelude.h:1682]: These functions are used as the implementations of built-ins in hlsl.meta.slang. Slang's `sign()` function returns `int`. Returning a float instead relies on implicit casting and could potentially cause downstream compiler warnings.
https://github.com/shader-slang/slang/blob/768ae5129f2c0bfd367e17f63ba4b8595f9024c8/source/slang/hlsl.meta.slang#L13784-L13793

**Comment by juliusikkala** [comment]
It's been such a long time since I started the initial version of this PR that I don't actually remember the specifics of why the test fails on WGSL, so I'll just ask the CI now. It's also possible that some of those problems have been fixed elsewhere in the meantime.

File: tests/hlsl-intrinsic/scalar-half.slang (hunk 4 above)
Comment (juliusikkala) [reply tests/hlsl-intrinsic/scalar-half.slang:11]: `asuint16` and `asfloat16` cannot be implemented due to WGSL missing 16-bit integer types: https://github.com/gpuweb/gpuweb/issues/273

Furthermore, implementations for many other built-ins in Slang depend on this capability, at least `isinf` and `isnan`, as WGSL also does not provide functions for those.

File: tests/hlsl-intrinsic/scalar-half.slang (hunk 3 above)
Comment (juliusikkala) [reply tests/hlsl-intrinsic/scalar-half.slang]: This appears to no longer be an issue with the newer spirv-opt, so I was able to re-enable optimizations on this test.

File: tests/hlsl-intrinsic/scalar-half.slang (hunk 2 above)
Comment (juliusikkala) [reply tests/hlsl-intrinsic/scalar-half.slang:3]: No, but I'm fairly certain that this is a bug on the DXC side, not ours.

I'm pretty sure it's this TODO in their constant folding:
https://github.com/microsoft/DirectXShaderCompiler/blob/b05d19d15d3ef97bfc2dcad3d47112f4f1a64e7a/tools/clang/lib/AST/ExprConstant.cpp#L3824

Comment (kaizhangNV) [reply tests/hlsl-intrinsic/scalar-half.slang:3]: OK, can we file an issue on our side, and put this link in the description and also reference this test?
I'm just afraid that after a while, nobody still remember there is test here still disabled.

The issue can be just a P3 issue, so we won't forget about it.

**Review by kaizhangNV — APPROVED** [review]
It looks good to me.
